llvm/lib/Transforms/Utils/LoopIterCountsToAssumptions.cpp

```diff
@@ -0,0 +1,215 @@
+    return nullptr;
+  }
+
+  if (S0) {
+    IVOperand = 0;
+    return S0;
+  }
+
+  if (S1) {
+    IVOperand = 1;
+    return S1;
+  }
+
+  return nullptr;
+}
+
+bool needStartValue(const Value &StartValue, const Value &MinIterValue) {
+  auto *I = dyn_cast<Instruction>(&MinIterValue);
+  if (!I)
+    return true;
+
+  for (Value *Op : I->operands()) {
+    if (Op == &StartValue)
+      return false;
+  }
+  return true;
+}
+
+} // namespace
+
+PreservedAnalyses
+LoopIterCountsToAssumptions::run(Loop &CurrentLoop, LoopAnalysisManager &AM,
+                                 LoopStandardAnalysisResults &AR,
+                                 LPMUpdater &U) {
+  ScalarEvolution *SE = &AR.SE;
+  AssumptionCache *AC = &AR.AC;
+  DominatorTree *DT = &AR.DT;
+
+  std::optional<int> RawMinIterationCount = getMinTripCount(&CurrentLoop);
+
+  // dump loop summary
+  LLVM_DEBUG(if (CurrentLoop.getLoopPreheader()) {
+    dbgs() << "Preheader:" << CurrentLoop.getLoopPreheader()->getName() << "\n";
+  });
+  LLVM_DEBUG(dbgs() << "Header:" << CurrentLoop.getHeader()->getName() << "\n");
+
+  if (!RawMinIterationCount.has_value() || RawMinIterationCount.value() <= 0)
+    return PreservedAnalyses::all();
+  int MinIterCount = RawMinIterationCount.value();
+
+  if (!isLoopRotateCandidate(CurrentLoop)) {
+    LLVM_DEBUG(
+        dbgs()
+        << "LoopIterCountsToAssumptions-Warning: Loop "
+        << CurrentLoop.getHeader()->getParent()->getName() << " "
+        << CurrentLoop.getName()
+        << " will not be rotated, therefore aborting adding assumptions.\n");
+    return PreservedAnalyses::all();
+  }
+
+  LLVM_DEBUG(CurrentLoop.getHeader()->getParent()->dump(););
+  LLVM_DEBUG(dbgs() << "Processing Loop Iteration Count Metadata: "
+                    << CurrentLoop.getHeader()->getParent()->getName() << " "
+                    << CurrentLoop.getName() << " (" << MinIterCount << ")\n");
+  auto *ExitBB = CurrentLoop.getExitingBlock();
+  if (!ExitBB)
+    return PreservedAnalyses::all();
+
+  // check if the branch instruction is the correct one
+  BranchInst *BI = dyn_cast<BranchInst>(ExitBB->getTerminator());
+  if (!BI || BI->isUnconditional())
+    return PreservedAnalyses::all();
+
+  ICmpInst *LoopCmpInstr = dyn_cast<ICmpInst>(BI->getCondition());
+  if (!LoopCmpInstr)
+    return PreservedAnalyses::all();
+
+  LLVM_DEBUG(dbgs() << "Branch Instruction Found: "; LoopCmpInstr->dump();
+             dbgs() << "\n");
+
+  if (LoopCmpInstr->getPredicate() == CmpInst::ICMP_EQ ||
+      LoopCmpInstr->getPredicate() == CmpInst::ICMP_NE) {
+    LLVM_DEBUG(dbgs() << "LoopIterCountsToAssumptions-Warning: EQ and NE "
+                         "conditions are currently not supported!");
+    return PreservedAnalyses::all();
+  };
+
+  // get Scalar Evolution of the induction variable
+  unsigned IVOperand = -1;
+  const SCEVAddRecExpr *AddRecExpr =
+      getLoopCmpIVAddRecExpr(*LoopCmpInstr, SE, IVOperand);
```

Comment:
I'd rather avoid "returning by ref", as this can be quite error prone, and it's not a clear interface. How about using a `pair` or `struct` as return type?

Then you can have something like `auto &[AddRecExpr, IVValue] = getLoopCmpIVAddRecExpr(*LoopCmpInstr, SE);`.

Or something like
```
struct IVInfo {
const SCEVAddRecExpr *AddRecExpr;
const Value *IVValue;
};
```
and have `getLoopCmpIVAddRecExpr` return an `std::optional<IVInfo>`